### DISCUSSION TITLE
does generateSitemaps create sitemap-index.xml ?

### DISCUSSION THREAD

**Discussion by jjang16** [vercel/next.js#61448] [App Router]
Hi, I'm planning to migrate my sitemap generating part of my project to next.js's feature by dynamically create multiple sitemaps using https://nextjs.org/docs/app/api-reference/file-conventions/metadata/sitemap#generating-multiple-sitemaps .

If sitemaps are dynamically created, does next.js have the ability to create the sitemap-index.xml ? otherwise the dynamically created sitemaps won't have any links pointing to them, making them unable to be discovered by search engines.

Thank you.

**Comment by spirituelle** [comment]
Apparently it doesn't and we really need it. also we need the dinamic id to be as string so we can name our ssitemap files as we want

**Reply by michaeldegori** [reply]
Yeah, this is wild. Documentation is really lacking on their newer sitemap generation

**Reply by spirituelle** [reply]
yeah. for now i just work with route.ts file to manually generate sitemap

**Reply by michaeldegori** [reply]
Also, have you figured out a way to generate sitemaps for pages with dynamic root params? Ie in my app/[city]/events, I can't figure out a good way to generate without mapping every single city and then fetching every single event which would be hundreds of thousands of entries. But is this how it should be done?

**Reply by michaeldegori** [reply]
I would obviously like the sitemap at /las-vegas/events to only have the events for las vegas listed

**Reply by michaeldegori** [reply]
@leerob, do you have any insights here?

**Reply by spirituelle** [reply]
inside the [city] folder you can create a route.ts file to generate your sitemap. and you can access to the city param with something like that
if (request.nextUrl.pathname.startsWith('/las-vegas')) {
    // fetch event for the city
  }
i am unsure if this is the best way to do it but i guess it will work out untill new developped versions of sitemaps appears

**Reply by michaeldegori** [reply]
Eesh, this makes sense. Thank you for the tip. I may just keep my existing (albeit ugly) sitemap setup in pages until docs and functionallity are improved in this area

**Reply by JannikZed** [reply]
@michaeldegori we are facing the same issue now, with a page with thousand of city + tour + language combinations. Did you find a nice way? The documentations are just really bad around that.

**Comment by tarun2324** [comment]
In Next.js 15, there are a couple of improvements that help with this use case:

- You can now use a string as the dynamic id in the generateSitemaps function, which lets you name your sitemap files as you want.
- For generating a sitemap index, you can create a route.ts file in your /sitemap-index.xml/ (can't name it sitemap | sitemap.xml) folder. This route can manually output an index of all your dynamically generated sitemap URLs (just like the generateSitemaps API does internally).

For large numbers of URLs or highly dynamic content, it’s actually better to skip the generateSitemaps function entirely. Instead, you can:

Define your sitemap index explicitly in route.ts. This way you can create multiple index sitemap. 

Pass your custom string id to the sitemap.ts function to generate the corresponding URL set on demand.

This approach avoids doing all the heavy lifting at build time and keeps things performant, especially if your sitemap changes often.

```
// app/sitemap-index.xml/route.ts
import { NextResponse } from 'next/server';

// Imagine this comes from your database or API
async function getSitemapIds(): Promise<string[]> {
  // Return identifiers like dates, categories, etc.
  return ['2025-07-17', '2025-07-16', '2025-07-15'];
}

// This route serves the sitemap index
export async function GET() {
  const baseUrl = `https://example.com/sitemap/`;

  const ids = await getSitemapIds();

  const sitemapsXml = ids.map(id => `
    <sitemap>
      <loc>${baseUrl}${id}.xml</loc>
    </sitemap>
  `).join('');

  const xml = `<?xml version="1.0" encoding="UTF-8"?>
<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
${sitemapsXml}
</sitemapindex>`;

  return new NextResponse(xml, {
    headers: { 'Content-Type': 'application/xml' },
  });
}
```

```
// app/sitemap.ts
import { MetadataRoute } from 'next';

// Replace this with your own data fetching logic
async function getUrlsForId(id: string): Promise<{ url: string; lastModified?: string }[]> {
  // Example: fetch all items for a given `id` (e.g., date or category)
  return [
    { url: `https://example.com/page/${id}/1`, lastModified: new Date().toISOString() },
    { url: `https://example.com/page/${id}/2`, lastModified: new Date().toISOString() },
  ];
}

export default async function sitemap({ id }: { id: string }): Promise<MetadataRoute.Sitemap> {
  const entries = await getUrlsForId(id);

  return entries.map(entry => ({
    url: entry.url,
    lastModified: entry.lastModified,
    changeFrequency: 'daily',
    priority: 0.7,
  }));
}

```

**Comment by BarnabasSzabolcs** [comment]
According to the official guide, sitemap-index.xml would be generated through export async function generateSitemaps but sadly it gives me 404.
